### PR TITLE
Feature: THREE.ShadowMapViewer

### PR DESCRIPTION
**Pull request by arya-s**
Hey,

as discussed in #5896, I created a simple utility to visualise a light's shadow map.

### CODE REVIEW

File: examples/js/utils/ShadowMapViewer.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+	//Holds the initial position and dimension of the HUD
+	var frame = { 
+		x: 10,
+		y: 10,
+		width: 256,
+		height: 256,
+	};
+
+	var camera = new THREE.OrthographicCamera( window.innerWidth / - 2, window.innerWidth / 2, window.innerHeight / 2, window.innerHeight / - 2, -1, 1 );
```
Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: To prevent user confusion, make the `near` plane positive so the frustum is in front of the camera. You will have to move the camera back a bit.

File: examples/js/utils/ShadowMapViewer.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		if ( this.enabled ) {
+
+			//Because a light's .shadowMap is only initialised after the first render pass
+			//we have to make sure the correct map is sent into the shader, otherwise we
+			//always end up with the scene's first added shadow casting light's shadowMap
+			//in the shader
+			//See: https://github.com/mrdoob/three.js/issues/5932
+			uniforms.tDiffuse.value = light.shadowMap;
+
+			renderer.autoClear = false; // To allow render overlay
```
Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: It is probably not a good idea to change `autoClear` behind-the-back of the user. Please be sure to reset it to its prior value.

File: examples/js/utils/ShadowMapViewer.js (hunk 1 above)
Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: Thank you, I updated the code with your suggestions.

File: examples/js/utils/ShadowMapViewer.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	var shader = THREE.UnpackDepthRGBAShader;
+
+	var uniforms = new THREE.UniformsUtils.clone( shader.uniforms );
+	var material = new THREE.ShaderMaterial( {
+					uniforms: uniforms,
+					vertexShader: shader.vertexShader,
+					fragmentShader: shader.fragmentShader
+	} );
+	var plane = new THREE.PlaneBufferGeometry( frame.width, frame.height );
+	var mesh = new THREE.Mesh( plane, material );
```
Comment (bchevalier) [inline examples/js/utils/ShadowMapViewer.js]: Hi,
I was wondering, as what you are trying to implement is to render 2D textures (with no perspective) on the screen, could it be simpler to use the ```THREE.Sprite``` component instead of a ```THREE.Mesh```?

There is an example of such an implementation here: [webgl sprites in Threejs](https://github.com/mrdoob/three.js/blob/master/examples/webgl_sprites.html)

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: Hey,

that was actually the first thing I looked into. However, a light's shadowMap is a WebGLRenderTarget and THREE.Texture doesn't accept those. And a THREE.SpriteMaterial doesn't accept that either.

Finally, I think you actually have to render out the texture before you can access it.

Correct me if I'm wrong and missed something.

Comment (bchevalier) [inline examples/js/utils/ShadowMapViewer.js]: I see. So basically, if we wanted this to work using ```THREE.Sprite```, ```THREE.SpriteMaterial ``` would have to accept ```THREE.WebGLRenderTarget```. I am not sure if it is feasible, but if it was done it could simplify a tiny bit the work of anyone building such HUDs.

Comment (mrdoob) [inline examples/js/utils/ShadowMapViewer.js]: I wonder that too. I haven't looked at that code for a bit. Would it be hard to make `SpriteMaterial` support `WebGLRenderTarget`?

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: @mrdoob I haven't looked into it. But in case of ```light.shadowMap```, you have to render it first. Sprites and Textures don't seem to require such a pass.

I'm not sure about other WebGLRenderTargets though.

Comment (mrdoob) [inline examples/js/utils/ShadowMapViewer.js]: > Edit: Maybe we should open a different issue and discuss that there?

Agreed!

File: examples/js/utils/ShadowMapViewer.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	//Holds the initial position and dimension of the HUD
+	var frame = { 
+		x: 10,
+		y: 10,
+		width: 256,
+		height: 256,
+	};
+
+	var camera = new THREE.OrthographicCamera( window.innerWidth / - 2, window.innerWidth / 2, window.innerHeight / 2, window.innerHeight / - 2, 1, 10 );
+	camera.position.set(0, 0, 1);
```
Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: This is dangerous. The near plane passes through the world origin. You are placing the mesh at the origin.

    camera.position.set( 0, 0, 2 );

Also, consider checking your formatting. http://zz85.github.io/mrdoobapproves/

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: @WestLangley thanks, I overlooked that.

**Comment by mrdoob** [comment]
Thanks!

File: examples/js/utils/ShadowMapViewer.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	this.render = function ( renderer ) {
+
+		if ( this.enabled ) {
+
+			//Because a light's .shadowMap is only initialised after the first render pass
+			//we have to make sure the correct map is sent into the shader, otherwise we
+			//always end up with the scene's first added shadow casting light's shadowMap
+			//in the shader
+			//See: https://github.com/mrdoob/three.js/issues/5932
+			uniforms.tDiffuse.value = light.shadowMap;
```
Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: Hmmm. `light` is defined???. This does appear to work, but is it guaranteed to?

`light` is passed into the constructor as an argument only.

File: examples/webgl_shadowmapviewer.html [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        dirLight = new THREE.DirectionalLight( 0xffffff, 1 );
+        dirLight.position.set( 0, 10, 0 );
+        dirLight.castShadow = true;
+        dirLight.shadowCameraNear = 0.01;
+        dirLight.shadowCameraFar = 10;
+        dirLight.shadowCameraRight = 15;
+        dirLight.shadowCameraLeft = -15;
+        dirLight.shadowCameraTop  = 15;
+        dirLight.shadowCameraBottom = -15;
+        spotLight.shadowDarkness = 0.5;
```
Comment (WestLangley) [inline examples/webgl_shadowmapviewer.html]: Should be `dirLight`.

File: examples/js/utils/ShadowMapViewer.js (hunk 5 above)
Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: What do you mean by guaranteed to work?

If ```light``` is undefined, the whole thing fails. Which is the same behaviour as for example ```THREE.DirectionalLightHelper```.

File: examples/webgl_shadowmapviewer.html (hunk 6 above)
Comment (arya-s) [inline examples/webgl_shadowmapviewer.html]: Thanks.

File: examples/js/utils/ShadowMapViewer.js (hunk 5 above)
Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: 1. Where is the variable `light` set?
2. Is is guaranteed to always be defined?

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: 1. Wherever the user defines his light setup, then passed in as an argument to the constructor only.
2. No.

I'm not sure about your concern here. IF ```light``` becomes undefined, the entire thing breaks down, similar to other utilities in three.js. Should the user be just warned and the application not break?

What would you suggest here?

Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: Sorry, I must not be explaining myself well. I'll try again. : - )

Look at `SpotLightHelper`:

    THREE.SpotLightHelper = function ( light ) {

	    THREE.Object3D.call( this );

	    this.light = light;

	    ...

	}

Note that `light` is retained as the property `this.light`.

Then, in `THREE.SpotLightHelper.prototype.update()`, it references `this.light`.

You do not do that.

Instead, you reference `light` in your `render()` method. Nowhere in your `render()` method do you define what `light` is.

It is as if a closure has been created -- which I expect it has, because it is working.

/ping @mrdoob -- who might be able to shed some light on this.

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: Well the constructor function encloses everything, and that's where ```light``` is an argument, so ```render()``` has access to that same argument.

Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: Yes. And your `render()` method is not part of the prototype. That is the difference.

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: I didn't put the methods on the prototype because I didn't want to expose internals of ```THREE.ShadowMapViewer``` such as the ```camera```, ```scene```, ```light```, etc. .

But I'm still not getting your point. Are you wondering why ```render()``` has access to ```light```, or are you concerned about every instance of ```THREE.ShadowMapViewer``` having their own ```render()```, ```update()```, etc.?

Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: Sorry, I should have noticed the subtle differences in your approach, and avoided stirring the pot. It is clear to me now.

It would be nice if a user could create one Viewer and use it to render any light's shadow map, though. But, that is not that big of a deal.

Thank you for your contribution!

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: Yea, that's an interesting approach. It would also allow some easier placing of the shadow maps (for example in a grid). Less overhead too (just one scene, one camera, many meshes).

I just tried to go about this like a light helper, one per object, but I guess I mixed things up with not using prototypes and such.

Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: What I meant was, expose `this.light`. That way, a user can instantiate one Viewer, and re-assign a different light. It is up to you.

Comment (arya-s) [inline examples/js/utils/ShadowMapViewer.js]: Ahhh, that's a good suggestion.

What's the procedure here? Do I update and do another pull request or should someone with rights just patch this in?

File: examples/webgl_shadowmapviewer.html (hunk 6 above)
Comment (mrdoob) [inline examples/webgl_shadowmapviewer.html]: Already fixed :)

File: examples/js/utils/ShadowMapViewer.js (hunk 5 above)
Comment (mrdoob) [inline examples/js/utils/ShadowMapViewer.js]: I think I would keep the current approach.

@WestLangley as far as I remember, we used `this.light` because there was no other way. I prefer the approach here whenever possible. Even if limiting the usage to `new THREE.ShadowMapViewer( ... )` may be less memory efficient than other approaches it also reduces the code paths we need to be aware of.

File: examples/js/utils/ShadowMapViewer.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+ * Example usage:
+ *	1) Include <script src='examples/js/utils/ShadowMapViewer.js'><script> in your html file
+ *
+ *	2) Create a shadow casting light and name it optionally:
+ *		var light = new THREE.DirectionalLight( 0xffffff, 1 );
+ *		light.castShadow = true;
+ *		light.name = 'Sun';
+ *
+ *	3) Create a shadow map viewer for that light and set its size and position optionally:
+ *		var shadowMapViewer = THREE.ShadowMapViewer( light );
```
Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: new THREE.ShadowMapViewer( light );

File: examples/js/utils/ShadowMapViewer.js (hunk 5 above)
Comment (WestLangley) [inline examples/js/utils/ShadowMapViewer.js]: @mrdoob OK.

@arya-s Almost always, the contributor makes a new pull request. Even if it is simple.
